src/modules/validator.js

```diff
@@ -85,56 +88,63 @@ class Validator {
-        if (valid) {
-            // Emit events about keys discovered
-            this.chluIpfs.events.emit('vendor pubkey', vMultihash);
-            this.chluIpfs.events.emit('vendor-marketplace pubkey', vmMultihash);
-            this.chluIpfs.events.emit('marketplace pubkey', mMultihash);
+        if (!useCache || !this.chluIpfs.cache.isValidityCached(hash)) {
+            const vmMultihash = this.keyLocationToKeyMultihash(popr.key_location);
+            const isExpectedKey = expectedPoPRPublicKey === null || expectedPoPRPublicKey === vmMultihash;
+            if (!isExpectedKey) {
+                throw new Error('Expected PoPR to be signed by ' + expectedPoPRPublicKey + ' but found ' + vmMultihash);
+            }
+            const mSignature = popr.marketplace_signature;
+            const vSignature = popr.vendor_signature;
+            const vMultihash = this.keyLocationToKeyMultihash(popr.vendor_key_location);
+            const vmSignature = popr.signature;
+            const marketplaceUrl = popr.marketplace_url;
+            const mMultihash = await this.fetchMarketplaceKey(marketplaceUrl, useCache);
+            const c = this.chluIpfs.crypto;
+            const validations = await Promise.all([
+                c.verifyMultihash(vMultihash, vmMultihash, vSignature),
+                c.verifyMultihash(mMultihash, vmMultihash, mSignature),
+                c.verifyMultihash(vmMultihash, hash, vmSignature)
+            ]);
+            // false if any validation is false
+            const valid = validations.reduce((acc, v) => acc && v);
+            if (valid) {
+                this.chluIpfs.cache.cacheValidity(hash);
+                // Emit events about keys discovered
+                this.chluIpfs.events.emit('vendor pubkey', vMultihash);
+                this.chluIpfs.events.emit('vendor-marketplace pubkey', vmMultihash);
+                this.chluIpfs.events.emit('marketplace pubkey', mMultihash);
+            } else {
+                throw new Error('The PoPR is not correctly signed');
+            }
+            return valid;
         } else {
-            throw new Error('The PoPR is not correctly signed');
+            return true;
         }
-        return valid;
     }
 
     keyLocationToKeyMultihash(l) {
         if (l.indexOf('/ipfs/') === 0) return l.substring('/ipfs/'.length);
         return l;
     }
 
-    async fetchMarketplaceKey(marketplaceUrl) {
+    async fetchMarketplaceKey(marketplaceUrl, useCache = true) {
         try {
             this.chluIpfs.logger.debug('Fetching marketplace key for ' + marketplaceUrl);
-            const response = await axios.get(marketplaceUrl + '/.well-known');
-            if (response.status !== 200) {
-                throw new Error('Expected HTTP Status Code 200, got ' + response.status + ' instead');
+            if (useCache) {
+                const multihash = this.chluIpfs.cache.getMarketplacePubKeyMultihash(marketplaceUrl);
```

Comment:
With the current way I organized ChluIPFS modules, having a `marketplace` module would mean that it contains logic used _by_ marketplaces only. So, following the current way things are organized, it makes sense that the function to get the marketplace public key is in the validator module because it's only necessary for validation.

Another options is having cache-wrapped functions but I don't like this since it takes away some control, for example I don't think it's a good idea to cache failures since they might be due to temporary unavailability of the marketplace or bugs.

I think for now it's better to keep it like this, maybe we can refactor it if we start using the cache in more places or if we change how we organize the modules